lib/recipe/bloc/recipe_bloc.dart

```diff
@@ -825,4 +853,163 @@ class RecipeBloc extends Bloc<RecipeEvent, RecipeState> {
       }
     }
   }
+
+  Recipe? _recipeOnDevice(recipeNumber) {
+    List<Recipe> deviceRecipes = List<Recipe>.from(state.recipes);
+    for (Recipe recipe in deviceRecipes) {
+      if (recipe.recipeNumber == recipeNumber) {
+        return recipe;
+      }
+    }
+    return null;
+  }
+
+  Future<void> _onRecipeImported(
```

Comment:
Apart from that, I think that this should be (possibly in the future) refactored to having more sub-functions with a higher abstraction level than the current low-level implementation of many parts of this function. Also some parts are quite a little tricky to understand, so having well-named subfunctions would help with that as well. This could be a refactoring `TODO` comment for now on, if we decide to leave it for later and get it merged now.